bankapp/src/main/kotlin/com/clouway/bankapp/adapter/spark/RegisterController.kt

```diff
@@ -0,0 +1,37 @@
+package com.clouway.bankapp.adapter.spark
+
+import com.clouway.bankapp.core.JsonSerializer
+import com.clouway.bankapp.core.UserAlreadyExistsException
+import com.clouway.bankapp.core.UserRegistrationRequest
+import com.clouway.bankapp.core.UserRepository
+import org.eclipse.jetty.http.HttpStatus
+import spark.Request
+import spark.Response
+import java.beans.PropertyChangeListener
+import java.beans.PropertyChangeSupport
+
+/**
+ * @author Tsvetozar Bonev (tsbonev@gmail.com)
+ */
+class RegisterController(private val userRepo: UserRepository,
+                         private val transformer: JsonSerializer) : Controller {
+
+    private val support : PropertyChangeSupport by lazy { PropertyChangeSupport(this) }
+
+    fun addPropertyChangeListener(plc: PropertyChangeListener){
+        support.addPropertyChangeListener(plc)
+    }
+
+    override fun handle(request: Request, response: Response): Any {
+        return try{
+            val user = userRepo
+                    .registerIfNotExists(
+                            transformer.fromJson(request.body(),
+                                    UserRegistrationRequest::class.java))
+            support.firePropertyChange("registration", null, user.username)
```

Comment:
userChangeListener.onRegistration()

listeners.forEach{it.onRegisration(user)}

listeners.forEach{it.onLogout(userId,userEmail)}



UserAsyncEventPublisher : UserChangeListener
UserChangeLogger:

client
asyncEventBus.publish(UserRegisteredEvent(userId, userName , email... )) 


microservice : 
/_ah/pubsubs/message -  PubSubsMessageHandler

messageEventBus.register(UserRegisrationHandler(), UserRegisteredEvent::class.java)

messageEventBus.register(UserLogaoutEventHandler(), UserLogaoutEvent::class.java)

Reply:
Changes applied.